cs3api4lab/auth/authenticator.py

```diff
@@ -1,31 +1,100 @@
+import importlib
+from datetime import datetime
+
+import jwt
 import cs3.gateway.v1beta1.gateway_api_pb2 as cs3gw
-import time
 import cs3.gateway.v1beta1.gateway_api_pb2_grpc as cs3gw_grpc
+
 from cs3api4lab.config.config_manager import Cs3ConfigManager
+import cs3.rpc.code_pb2 as cs3code
+from tornado import web
+
+import cs3.gateway.v1beta1.gateway_api_pb2 as cs3gw
+import cs3.gateway.v1beta1.gateway_api_pb2_grpc as cs3gw_grpc
 from cs3api4lab.auth.channel_connector import ChannelConnector
 
 
 class Authenticator:
-    tokens = {}  # map client_id [string] to {authentication token, token expiration time}
+    tokens = {}
+    cs3_stub = None
 
-    def __init__(self, config, channel):
+    def __init__(self, config=None, log=None):
         self.config = config
-        self.cs3_api = cs3gw_grpc.GatewayAPIStub(channel)
+        self.log = None
+        channel = ChannelConnector().get_channel()
+        self.cs3_stub = cs3gw_grpc.GatewayAPIStub(channel)
+
+    """
+    Parent class for different authentication types. Authentication class must implement the authenticate method.
+    If authentication class can refresh token must implement the refresh token method.
+    Using auth type is declared in the config file.
+    """
 
     def authenticate(self, client_id):
-        # ToDo: use real authentication data or get token from author provider
-        # authReq = cs3gw.AuthenticateRequest(type='bearer', client_secret=userid)
-
-        if client_id not in self.tokens or self.tokens[client_id]['exp'] < time.time():
-            auth_req = cs3gw.AuthenticateRequest(type=self.config['login_type'],
-                                                 client_id=self.config['client_id'],
-                                                 client_secret=self.config['client_secret'])
-            auth_res = self.cs3_api.Authenticate(auth_req)
-            self.tokens[client_id] = {'tok': auth_res.token,
-                                    'exp': time.time() + float(self.config['auth_token_validity'])}
-
-        # piggy back on the opportunity to expire old tokens, but asynchronously
-        # as to not impact the current session: let's use python3.7's coroutines support
-        # asyncio.run(_async_cleanuptokens())
-
-        return self.tokens[client_id]['tok']
+        """
+        The basic authenticate method, return IOP token or refresh if is not present or invalid.
+        """
+        if client_id not in self.tokens or (client_id in self.tokens and self._check_token(self.tokens[client_id])):
+            self.refresh_token(client_id)
```

Comment:
Why do you keep passing `client_id` all over the place if this is part of `self.config`?

Reply:
I don't remember "Why". In the first code version, I'd like to gets client_id from get/post request. I change it.

Reply:
I think your initial idea was that this could be made a library in the future and used by other clients that might authenticate with different users. But I'm still not 100% this will be needed... In lab for sure it's not, so I would agree with a clean up on this.

Reply:
👌